oshmem: Reduce the service memory footprint

sb17v opened the pull request:
Current implementation is maintaining a static cache of 32 entries
per remote peer containing memory key information.
In practice, only 2-3 entries are used. At scale, however this leads
to a significant overhead.

Example:
* one cache entry is 40B
* with default 32 entries per peer it is 1280B per peer.
* on cluster with 1000 nodes and 40 PPN it results in 51MB per process
* Cumulatively, per node (given 40PPN) its 51MB x 40 = 2GB of overhead

This commit introduces dynamic cache that contains only required number
of entries. With 2 entries by default this leads to 16x reduction in the
cache memory footprint.

Future optimization of the cache entry layout is planned.
Will be PR'd as a follow-up.

Co-authored-by: Artem Y. Polyakov <artemp@nvidia.com>
Signed-off-by: Subhadeep Bhattacharya <subhadeepb@nvidia.com>
Signed-off-by: Artem Polyakov <artemp@nvidia.com>

ggouaillardet committed:
oshmem: add pkg-config files

Refs open-mpi/ompi#4730

Remove -lopen-rte reference

Signed-off-by: Gilles Gouaillardet <gilles@rist.or.jp>

manjugv committed:
Changes to env var name and behavior to reflect the OpenSHMEM spec

Signed-off-by: Manjunath Gorentla Venkata <manjunath@nvidia.com>

awlauria committed:
Remove residual references to -lopen-rte

Signed-off-by: Austen Lauria <awlauria@us.ibm.com>

awlauria committed:
Cleanup 'orte' ompi dir.

Signed-off-by: Austen Lauria <awlauria@us.ibm.com>

awlauria committed:
Cleanup 'orte' opal dir.

Signed-off-by: Austen Lauria <awlauria@us.ibm.com>

awlauria committed:
Rework things to replace hard-coded '-lopen-pal' with @OPAL_LIB_NAME@.

This makes it consistent with OMPI_LIBMPI_NAME.

Replace/remove OPAL_LIB_PREFIX since it currently is unused/dead code.

Signed-off-by: Austen Lauria <awlauria@us.ibm.com>

awlauria committed:
Removed out-dated calls to opal_progress_event_users_*() from init/finalize.

Since the event library has its own thread, trying reduce latency doesn't
make much sense anymore if there's no consumers in this case.

Signed-off-by: Austen Lauria <awlauria@us.ibm.com>

Commit:
Add support for UCC collectives to OSHMEM

Signed-off-by: Ferrol Aderholdt <faderholdt@nvidia.com>

tkordenbrock committed:
portals4: eliminate header order dependencies in the Portals4 components

Signed-off-by: Todd Kordenbrock <thkgcode@gmail.com>

sb17v committed:
oshmem: Reduce the service memory footprint

Current implementation is maintaining a static cache of 32 entries
per remote peer containing memory key information.
In practice, only 2-3 entries are used. At scale, however this leads
to a significant overhead.

Example:
* one cache entry is 40B
* with default 32 entries per peer it is 1280B per peer.
* on cluster with 1000 nodes and 40 PPN it results in 51MB per process
* Cumulatively, per node (given 40PPN) its 51MB x 40 = 2GB of overhead

This commit introduces dynamic cache that contains only required number
of entries. With 2 entries by default this leads to 16x reduction in the
cache memory footprint.

Future optimization of the cache entry layout is planned.
Will be PR'd as a follow-up.

Co-authored-by: Artem Y. Polyakov <artemp@nvidia.com>
Signed-off-by: Subhadeep Bhattacharya <subhadeepb@nvidia.com>
Signed-off-by: Artem Polyakov <artemp@nvidia.com>

Commit:
Add Cornelis Networks to the LICENSE file.

Signed-off-by: Michael Heinz <mheinz@cornelisnetworks.com>

Commit:
Merge pull request #9043 from mwheinz/mheinz-9042-master

Add Cornelis Networks to the LICENSE file.

tkordenbrock committed:
Merge pull request #9007 from tkordenbrock/portals4/master/remove-include-ordering-dependency

master: portals4: eliminate header order dependencies in the Portals4 components

sb17v commented:
@janjust @karasevb FYI

karasevb commented on oshmem/mca/spml/ucx/spml_ucx.c:
```diff
@@ -104,6 +104,155 @@ int mca_spml_ucx_enable(bool enable)
+                                    &((*ucx_mkey)->rkey));
+        if (UCS_OK != err) {
+            SPML_UCX_ERROR("failed to unpack rkey: %s", ucs_status_string(err));
+            return rc;
```
Should return `OSHMEM_ERROR` instead of `rc`?

sb17v replied:
Yes. Fixed.

awlauria committed:
Merge pull request #9015 from wfaderhold21/topic/osh-ucc

OSHMEM/SCOLL: Add support for UCC collectives to OSHMEM

manjugv commented:
@sb17v @artemp Please make sure that you run correctness benchmarks - OSU, SOS, and vendor tests (https://github.com/openshmem-org/tests-mellanox).

artpol84 commented:
Thanks @manjugv
We did SOS, but will do OSU and the vendor one as well.

awlauria committed:
Merge pull request #7977 from manjugv/topic/oshmem_envvars

Change env var name and behavior to reflect the OpenSHMEM spec

awlauria committed:
Merge pull request #8976 from awlauria/cleanup_orte

pkg-config oshmem files and removing some -lopen-rte references

rhc54 committed:
Update PMIx/PRRTE pointers

Pickup fixes to node resolution

Signed-off-by: Ralph Castain <rhc@pmix.org>

rhc54 committed:
Merge pull request #9067 from rhc54/topic/fix

Update PMIx/PRRTE pointers

hjelmn committed:
btl/uct: add support for UCX 1.10.x

This commit adjusts for yet another API break in UCT. This time it is the removal of
a parameter to a callback function. This change should allow btl/uct to be used with
1.10 so bumping to allow version to reflect the change.

Signed-off-by: Nathan Hjelm <hjelmn@google.com>

hjelmn committed:
Merge pull request #9074 from hjelmn/btl_uct_1_10

btl/uct: add support for UCX 1.10.x

rhc54 committed:
ompi_setup_cxx.m4: Don't require a C++ compiler

Gracefully ignore the lack of a C++ compiler.  Specifically, do not
invoke AC_PROG_CXXCPP unless we actually found a C++ compiler.

Also remove a spurious check for $CXX to be empty (which, by the very
comment above it, will never be true).

Signed-off-by: Ralph Castain <rhc@pmix.org>
Signed-off-by: Jeff Squyres <jsquyres@cisco.com>

rhc54 committed:
Merge pull request #9075 from rhc54/topic/cpp

Don't require a C++ compiler

rhc54 committed:
Update PRRTE pointer

Include hostname fixes for #9068

Signed-off-by: Ralph Castain <rhc@pmix.org>

bwbarrett committed:
Merge pull request #9078 from rhc54/topic/up

Update PRRTE pointer

hpcraink committed:
If AMODE not set in MPI_File_open, not all of ompio_fh is set, fixes #9083.

Signed-off-by: Rainer Keller <rainer.keller@hs-esslingen.de>

hpcraink committed:
Add braces per Jeff's comment in #9084

Signed-off-by: Rainer Keller <rainer.keller@hs-esslingen.de>

jsquyres committed:
Merge pull request #9084 from hpcraink/master

If AMODE not set in MPI_File_open, not all of ompio_fh is set, fixes …

Commit:
fbtl-posix: link to common_ompio

The posix fbtl calls mca_common_ompio_progress(), which resides in
common/ompio (i.e., libmca_common_ompio.la).  So add that into
mca_fbtl_posix_la_LIBADD (like we do in a few other OMPIO-based
components).  Failure to do this *can* lead to the posix fbtl
component failing to load (depending on whether other OMPIO-based
components that pull in libmca_common_ompio were loaded first).

Thanks to Honggang Li for raising the issue.

Signed-off-by: Jeff Squyres <jsquyres@cisco.com>

jsquyres committed:
Merge pull request #9086 from jsquyres/pr/fbtl-posix-link-fix

fbtl-posix: link to common_ompio
